File: src/python/xerxes/operators/plans.py
# Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""Session-scoped owner of the operator's structured execution plan.

Wraps a single :class:`OperatorPlanState` so the runtime can hold a stable
reference even when the plan is replaced wholesale by the ``update_plan``
tool.
"""

from __future__ import annotations

from .types import OperatorPlanState


class PlanStateManager:
    """Stateful holder around one :class:`OperatorPlanState` instance.

    The manager exists so that the surrounding session has a stable object
    to point at — the underlying plan state can be mutated in-place by
    :meth:`update` without invalidating cached references.
    """

    def __init__(self) -> None:
        """Allocate an empty plan state."""

        self._state = OperatorPlanState()

    @property
    def state(self) -> OperatorPlanState:
        """The live plan state owned by this manager."""

        return self._state

    def update(self, explanation: str | None, plan: list[dict[str, str]]) -> dict:
        """Replace the plan contents and return the new wire snapshot.

        Args:
            explanation: New plan preamble (``None`` clears it).
            plan: Ordered list of ``{"step": ..., "status": ...}`` dicts.
        """

        return self._state.update(explanation, plan)

    def summary(self) -> str:
        """Return a short status string for the first few plan steps."""

        if not self._state.steps:
            return "No plan"
        return ", ".join(f"{step.status}:{step.step}" for step in self._state.steps[:3])

File: src/typescript/src/operators/plans.ts
// Copyright 2026 The Xerxes-Agents Author @erfanzar (Erfan Zare Chavoshi).
// Licensed under the Apache License, Version 2.0.

import type { OperatorPlanStateSnapshot, OperatorPlanStep } from './types.js'

export interface PlanStateManagerOptions {
  readonly now?: () => Date
}

/** Session-scoped structured plan with monotonic revisions. */
export class PlanStateManager {
  private explanation: string | undefined
  private revision = 0
  private steps: OperatorPlanStep[] = []
  private updatedAt: string
  private readonly now: () => Date

  constructor(options: PlanStateManagerOptions = {}) {
    this.now = options.now ?? (() => new Date())
    this.updatedAt = this.now().toISOString()
  }

  get state(): OperatorPlanStateSnapshot {
    return this.snapshot()
  }

  update(explanation: string | undefined, plan: readonly OperatorPlanStep[]): OperatorPlanStateSnapshot {
    this.explanation = normalizeExplanation(explanation)
    this.steps = plan.map(step => Object.freeze({
      step: requireStep(step.step),
      status: normalizeStatus(step.status),
    }))
    this.revision += 1
    this.updatedAt = this.now().toISOString()
    return this.snapshot()
  }

  summary(): string {
    if (!this.steps.length) return 'No plan'
    return this.steps.slice(0, 3).map(step => `${step.status}:${step.step}`).join(', ')
  }

  private snapshot(): OperatorPlanStateSnapshot {
    return Object.freeze({
      ...(this.explanation === undefined ? {} : { explanation: this.explanation }),
      revision: this.revision,
      updatedAt: this.updatedAt,
      steps: Object.freeze(this.steps.map(step => Object.freeze({ ...step }))),
    })
  }
}

function normalizeExplanation(value: string | undefined): string | undefined {
  if (value === undefined) return undefined
  const trimmed = value.trim()
  return trimmed || undefined
}

function requireStep(value: string): string {
  const trimmed = value.trim()
  if (!trimmed) throw new TypeError('Plan step must not be empty')
  return trimmed
}

function normalizeStatus(value: string): string {
  const trimmed = value.trim()
  return trimmed || 'pending'
}
